numpy/core/src/umath/loops.c.src

```diff
@@ -2618,41 +2618,48 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
 NPY_NO_EXPORT void
 OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
 {
-#if defined(NPY_PY3K)
     PyObject *zero = PyLong_FromLong(0);
+
     UNARY_LOOP {
         PyObject *in1 = *(PyObject **)ip1;
         PyObject **out = (PyObject **)op1;
+        PyObject *ret = NULL;
         int v;
-        PyObject *ret;
-        if (PyObject_Cmp(in1 ? in1 : Py_None, zero, &v) == -1) {
-            return;
+
+        if (in1 == NULL) {
+            in1 = Py_None;
+        }
+
+        v = PyObject_RichCompareBool(in1, zero, Py_LT);
+        if (v == 1) {
+            ret = PyLong_FromLong(-1);
+        } else if (v == 0) {
+            v = PyObject_RichCompareBool(in1, zero, Py_GT);
+            if (v == 1) {
+                ret = PyLong_FromLong(1);
+            } else if (v == 0) {
+                v = PyObject_RichCompareBool(in1, zero, Py_EQ);
+                if (v == 1) {
+                    ret = PyLong_FromLong(0);
+                }
```

Comment:
@jaimefrio I have already implemented the above and its working as expected. For the error message, I was gonna ask you, what it should be :stuck_out_tongue: . 

How about we go with the error message you suggested in the mailing list or how about we just explicitly tell the user that the exception is due to NaN. Like  `TypeError: Trying to compare NaN` (Something like this).